### Question

Imagine you are already working as a Data Analyst intern and you have been asked to
perform some data analysis on 3 cities; Riyadh, Munich and Panama City. The goal is to help
your team to better understand the current situation in each city so that we can better plan and
execute strategies that will yield better results.

### Import Libraries

In [1]:
import numpy as np
import pandas as pd
import datetime

#Plotting
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Data Importation

In [2]:
df = pd.read_excel('Downloads/DiDi.xlsx', sheet_name='rawdata')

In [3]:
df.head()

driver id        driver_role driver_city_name  sign_up sign_up_date  \
0  8470311823362  Driver from fleet           Riyadh  Sign up   2020-12-16   
1  8470311842205  Driver from fleet      Panama City  Sign up   2020-12-09   
2  8470311842215  Driver from fleet           Munich  Sign up   2020-12-31   
3  8470311842386  Driver from fleet      Panama City  Sign up   2020-09-09   
4  8470311842388  Driver from fleet      Panama City  Sign up   2021-01-18   

       regular regular_date      bind_car bind_car_date      available  ...  \
0      Regular   2020-12-18      Bind car    2020-12-18      Available  ...   
1      Regular   2021-01-04      Bind car    2021-03-12      Available  ...   
2  Not regular          NaT      Bind car    2020-12-31  Not available  ...   
3      Regular   2020-09-15      Bind car    2021-01-21      Available  ...   
4  Not regular          NaT  Not bind car           NaT  Not available  ...   

  main_channel car_factory_year   brand_id  inn_id first_download_time  \
0      Organic           2018.0  1014091.0     NaN          2020-12-16   
1      Organic           2015.0  1014167.0     NaN          2020-09-03   
2      Organic              NaN        NaN     NaN                 NaT   
3      Organic           2014.0  1014493.0     NaN          2020-09-05   
4      Organic              NaN        NaN     NaN                 NaT   

  driver_first_trip_date driver_last_trip_date car_first_trip_date acceptance  \
0             2021-01-20            2021-01-22          2021-01-20       32.0   
1             2021-01-12            2021-04-08          2021-03-12      897.0   
2                    NaT                   NaT                 NaT        NaN   
3             2021-01-23            2021-04-08          2021-01-23     1399.0   
4                    NaT                   NaT                 NaT        NaN   

   trips  
0   27.0  
1  467.0  
2    NaN  
3  948.0  
4    NaN  

[5 rows x 21 columns]

In [4]:
df.shape

(15387, 21)

The DataFrame consists of **15387** rows and **21 columns**

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15387 entries, 0 to 15386
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   driver id               15387 non-null  int64         
 1   driver_role             15387 non-null  object        
 2   driver_city_name        15387 non-null  object        
 3   sign_up                 15387 non-null  object        
 4   sign_up_date            15387 non-null  datetime64[ns]
 5   regular                 15387 non-null  object        
 6   regular_date            9372 non-null   datetime64[ns]
 7   bind_car                15387 non-null  object        
 8   bind_car_date           9426 non-null   datetime64[ns]
 9   available               15387 non-null  object        
 10  available_date          7521 non-null   datetime64[ns]
 11  main_channel            15236 non-null  object        
 12  car_factory_year        8280 non-null   float6

In [6]:
numerical_features = [features for features in df.columns if df[features].dtype != 'O']
print('Number of Numerical Features:', len(numerical_features))

Number of Numerical Features: 13


In [7]:
categorical_features = [features for features in df.columns if df[features].dtype == 'O']
print('Number of Categorical Features:', len(categorical_features))

Number of Categorical Features: 8


In [8]:
features_with_na = [features for features in df.columns if df[features].isnull().sum() > 1]

for feature in features_with_na:
    print(feature, np.round(df[feature].isnull().mean(), 4), '% missing values')

regular_date 0.3909 % missing values
bind_car_date 0.3874 % missing values
available_date 0.5112 % missing values
main_channel 0.0098 % missing values
car_factory_year 0.4619 % missing values
brand_id 0.4619 % missing values
inn_id 0.9837 % missing values
first_download_time 0.4161 % missing values
driver_first_trip_date 0.5713 % missing values
driver_last_trip_date 0.5713 % missing values
car_first_trip_date 0.5953 % missing values
acceptance 0.5603 % missing values
trips 0.5603 % missing values


From above we see that there are missing values in some of the variables

**1. What is the Driver ID of the driver who has completed the most trips?**

In [9]:
most_completed_trips = df.groupby(['driver id'])['trips'].sum().sort_values(ascending=False)

In [11]:
most_completed_trips.head(1)

driver id
8470311935249    3313.0
Name: trips, dtype: float64

Driver ID: 8470311935249 completed the most trips

**2. Which city has the highest average trips per Driver?**

In [12]:
highest_average_trips = df.groupby(['driver_city_name','driver id'])['trips'].mean().sort_values(ascending=False)

In [13]:
highest_average_trips.head(1)

driver_city_name  driver id    
Panama City       8470311946332    2733.0
Name: trips, dtype: float64

Panama City has the highest average trips per Driver

**3. Which month had the 3rd lowest number of signups in Munich?**

In [14]:
new_df = df[df['driver_city_name'] == 'Munich']
new_df.head()

driver id        driver_role driver_city_name  sign_up sign_up_date  \
2   8470311842215  Driver from fleet           Munich  Sign up   2020-12-31   
5   8470311842445  Driver from fleet           Munich  Sign up   2020-12-17   
10  8470311842506  Driver from fleet           Munich  Sign up   2020-12-16   
13  8470311842550  Driver from fleet           Munich  Sign up   2020-09-16   
14  8470311842613          Driver-IE           Munich  Sign up   2020-09-19   

        regular regular_date      bind_car bind_car_date      available  ...  \
2   Not regular          NaT      Bind car    2020-12-31  Not available  ...   
5       Regular   2020-12-21      Bind car    2021-01-04      Available  ...   
10  Not regular          NaT  Not bind car           NaT  Not available  ...   
13      Regular   2020-12-16  Not bind car           NaT  Not available  ...   
14  Not regular          NaT      Bind car    2020-09-26  Not available  ...   

   main_channel car_factory_year   brand_id    inn_id first_download_time  \
2       Organic              NaN        NaN       NaN                 NaT   
5       Organic           2008.0  1014309.0       NaN          2020-09-09   
10      Organic              NaN        NaN       NaN                 NaT   
13      Organic              NaN        NaN       NaN          2020-12-16   
14      Organic           2011.0  1014091.0  Approved          2021-02-27   

   driver_first_trip_date driver_last_trip_date car_first_trip_date  \
2                     NaT                   NaT                 NaT   
5              2021-01-27            2021-04-13          2021-01-27   
10                    NaT                   NaT                 NaT   
13                    NaT                   NaT                 NaT   
14                    NaT                   NaT          2020-10-15   

   acceptance   trips  
2         NaN     NaN  
5      1543.0  1030.0  
10        NaN     NaN  
13        NaN     NaN  
14        NaN     NaN  

[5 rows x 21 columns]

In [15]:
new_df['sign_up_date'] = new_df['sign_up_date'].dt.month 

<ipython-input-15-e8654385dec5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['sign_up_date'] = new_df['sign_up_date'].dt.month


In [17]:
lowest_num_signups = new_df.groupby(['sign_up_date'])['sign_up'].count().sort_values(ascending=False)

In [18]:
pd.DataFrame(lowest_num_signups)

sign_up
sign_up_date         
1                1869
12               1582
2                1059
3                 553
4                 195
9                  13
10                  9
11                  5

Month 9 which is September, has the 3rd lowest number of signups